### PR TITLE
🔨 Update tools for development on Apple Silicon

### PR DESCRIPTION
**Pull request by chris-araman**
Ruby 2.4.3 is no longer in support and fails to compile with current tools. The `danger` and `xcpretty` gems seem to run successfully on Ruby 3.0.0.

The `swiftenv` formula is not available in the `homebrew/core` tap, so we now specify its tap as `kylef/formulae`, per instructions at https://swiftenv.fuller.li/.

An arm64 `shellcheck` bottle is not available because `ghc` is not yet available for arm64 via Homebrew (https://github.com/Homebrew/homebrew-core/pull/65997). I've added `shfmt` in the meantime, as it is complementary. I'll be happy to add `shellcheck` back once it's available for arm64.

Some of the scripts were indenting with 4 spaces, and others with 2. I decided to standardize on 2 for brevity, but am happy to change it if others disagree.

I thought it might be a good idea to install the Swift linters via Homebrew and remove our dependence on Mint. The linters weren't all available from Homebrew when first added to this project, but they are now. Mint is currently problematic on M1 Macs because it fails to link globally to `/usr/local/bin` when that path does not yet exist. Homebrew for M1 Macs defaults to `/opt/homebrew/bin`. One thing we lose here is the ability to pin to particular versions of the linters, as Homebrew only supports installing the "current" version. I think that's a fair trade-off.

The Homebrew bundle lockfile isn't a typical lockfile in that it doesn't support version pinning, however it does add some value in that it provides evidence of a last-known-good state. However, it doesn't seem to handle removed or renamed brews/taps/mas entries. To work around that oversight, I've started wiping it during mas build bootstrapping. This allows us to maintain a clearer picture of the last-known-good Homebrew bundle state. However, I'd be happy to remove this "lockfile" and start passing `--no-lock` instead, if others feel it isn't meeting our needs.

The `sort` scripts were used when CocoaSeeds were still in use, but references to both have previously been removed.

I've enabled the `swiftformat` and `swift-format` linters in addition to the existing `swiftlint` linter. Getting them to agree on certain rules took a bit of finesse.

### CODE REVIEW

**Comment by chris-araman** [comment]
> Some files could not be reviewed due to errors:
> Loading configuration from '.swiftlint.yml'
> Invalid configuration for 'opening_brace'. Falling back to default.
Hound supports only SwiftLint 0.27.0 from July 2018. The current release is 0.43.1 from last week.
http://help.houndci.com/en/articles/2461415-supported-linters
https://github.com/realm/SwiftLint/releases/tag/0.27.0
https://github.com/realm/SwiftLint/releases/tag/0.43.1

Should I try to make the configuration work for both versions, or should I disable SwiftLint in `.hound.yml`? Lint is already covered by `Jenkinsfile`.

**Comment by chris-araman** [comment]
> Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration. (opening_brace)

These appear to be a disagreement between Hound's older version of SwiftLint and the current version.

File: MasKit/AppStore/SSPurchase.swift [hunk 1]
```diff
@@ -21,7 +21,7 @@ extension SSPurchase {
             "price": 0,
             "salableAdamId": adamId,
             "pg": "default",
-            "appExtVrsId": 0
+            "appExtVrsId": 0,
```
[the remark this answers is not in the document]
Comment (chris-araman) [reply MasKit/AppStore/SSPurchase.swift:24]: Apple's `swift-format` enforces trailing commas in multiline collection initializers when more than one element is listed. I've configured `swiftformat` and `swiftlint` to expect the same. I've also disabled SwiftLint in Hound, as it only offers an older version.
https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8

**Comment by chris-araman** [comment]
@phatblat, is Jenkins healthy? It does look like builds have been successful there lately. I don't want to risk regressions to the Jenkins build, but I don't know what the healthy baseline should be.
https://jenkins.log-g.co/job/mas-cli/job/mas/view/change-requests/builds

File: MasKit/AppStore/SSPurchase.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (phatblat) [reply MasKit/AppStore/SSPurchase.swift:24]: Sounds good.

File: MasKitTests/Formatters/AppListFormatterSpec.swift [hunk 2]
```diff
@@ -25,14 +26,14 @@ class AppListsFormatterSpec: QuickSpec {
...
+                let product = SoftwareProductMock(
                     appName: "Awesome App",
                     bundleIdentifier: "",
                     bundlePath: "",
                     bundleVersion: "19.2.1",
                     itemIdentifier: 12345
-                    )]
-                let output = format(products)
+                )
+                let output = format([products])
```
Comment (phatblat) [inline MasKitTests/Formatters/AppListFormatterSpec.swift:36]: This should reference the new local `product` now.

```suggestion
                let output = format([product])
```

**Review by phatblat — COMMENTED** [review]
@chris-araman Sorry about Jenkins. The man builds have been stalling on either Bundler installing gems or Homebrew checking tools. I'll take a look at it today.

**Comment by phatblat** [comment]
Fantastic cleanup, @chris-araman ! 🧹💨🚿✨

I've been going back and forth on formatting tools. For a while I used `swiftlint --format`, but that did some really weird things when the code got complex. Since reading https://nshipster.com/swift-format/ I've been debating between using Nick' SwiftFormat or Apple's swift-format. Which tool did you use to format for this PR?

I don't think it's worth supporting two versions of SwiftLint so I'm fine with turning off Hound for now.

`shellcheck` was mainly for lint warnings in scripts. Does `shfmt` output any type of lint warnings? If not we can just refer to Hound's shellcheck PR comments in the meantime.

I'm very happy to see the change to 2 space indentation in scripts. A lot of the scripts started with 4 spaces but then I switched to [2 spaces](https://github.com/mas-cli/mas/blob/master/.editorconfig#L13) and never went back to update the other scripts.

Thanks for removing sort. CocoaSeeds was canned when I saw the tool was abandoned (never seen any project but this one use it). Also perl (ew!) 😷.

I like the idea of using the Homebrew bundle lockfile as mainly metadata for devs. It'll probably thrash a bit especially as I'm still on an x86 for the time being. Also, I'm glad to see switching away from Mint. That was sort of an experiment but really fouled things up with testing the Homebrew formula and building bottles as Homebrew would reinstall Carthage since the mas formula depends on it.

Regarding Carthage, I'm not a fan of checking in copies of dependencies' source but I've left it that way ever since #169  because that was such a nightmare with me trying to get CocoaPods to work before @tonyarnold rescued the switch away from the latent CocoaSeeds. Curious about your thoughts on ignoring the `Carthage/` dir since the scripts use the tool to bootstrap anyway.

**Comment by phatblat** [comment]
Jenkins is hanging on the [`brew bundle install`](https://github.com/chris-araman/mas/blob/build/script/bootstrap#L19) command. This is probably happening because it runs under a non-admin user. I guess this used to be flaky and only break when a new tool was being installed. Now I think it's trying to install all tools on each run and then hanging.

Can you try wrapping the homebrew stuff in `bootstrap` with the following check?

```bash
if [[ $USER != "jenkins" ]]; then
  rm ...
  brew ...
fi
```

**Comment by phatblat** [comment]
Ugh, jenkins is certainly **not** healthy. It skips over Homebrew in #338 but then ran forever building Carthage dependencies until it hit the 1h timeout. I'm having trouble figuring out what's going on as VNC and SSH are timing out. Guessing that the hard drive is wearing out. Planning to replace the box with an M1 mini very soon.

Let's just ignore Jenkins status for now. I will validate your changes manually.

File: MasKitTests/Formatters/AppListFormatterSpec.swift (hunk 2 above)
Comment (phatblat) [reply MasKitTests/Formatters/AppListFormatterSpec.swift]: I applied this change but then found that `AppListFormatter` isn't included in the Xcode build and it's addition to the list command got stepped on by a merge mistake. Fixed in #339

File: MasKitTests/Formatters/AppListFormatterSpec.swift [hunk 3]
```diff
@@ -33,7 +33,7 @@ class AppListsFormatterSpec: QuickSpec {
                     bundleVersion: "19.2.1",
                     itemIdentifier: 12345
                 )
-                let output = format([products])
+                let output = format([product])
```
Comment (chris-araman) [inline MasKitTests/Formatters/AppListFormatterSpec.swift:36]: 🤦🏻‍♂️ Good catch!

**Comment by chris-araman** [comment]
> I've been going back and forth on formatting tools. For a while I used `swiftlint --format`, but that did some really weird things when the code got complex. Since reading https://nshipster.com/swift-format/ I've been debating between using Nick' SwiftFormat or Apple's swift-format. Which tool did you use to format for this PR?

I ended up using all three, in an effort to get them to play nicely together. I'm not sure if it's worth the effort in the long run. Long term, I'd happily drop two to keep things simple. I'm not sure which I'd pick at this point, though. They're all still in active development. For the time being, they're each autoformatting and are run in a particular order.

> `shellcheck` was mainly for lint warnings in scripts. Does `shfmt` output any type of lint warnings? If not we can just refer to Hound's shellcheck PR comments in the meantime.

I'm not sure what the overlap between `shellcheck` and `shfmt` is. It seems like both are popular, and some projects use both. `shfmt -d` will emit a diff and return a non-zero exit code if it would have formatted the script differently, but it won't emit a formatted warning message like a linter would. As I wrote it, `shfmt` will just autoformat along with the three Swift linters.
 
> Regarding Carthage, I'm not a fan of checking in copies of dependencies' source but I've left it that way ever since #169 because that was such a nightmare with me trying to get CocoaPods to work before @tonyarnold rescued the switch away from the latent CocoaSeeds. Curious about your thoughts on ignoring the `Carthage/` dir since the scripts use the tool to bootstrap anyway.

I agree that the committed copies are awkward. I'm for removing them and adding them to `.gitignore`, or replacing them with submodules if Carthage supports that. Separate PR, though, I think.

**Review by phatblat — APPROVED** [review]

**Comment by phatblat** [comment]
> Maybe I should change script/lint to autoformat (for developer convenience) and also to lint and fail (for CI)?

I would like to have a check fail when the linter returns errors. Could you split out the formatters into their own script? Then `lint` can run in a GitHub Action and `format` (or whatever it's called) can just be run manually by devs.

**Comment by chris-araman** [comment]
Sounds good. I'll split out `lint` and `format` in some additional commits.

I've rebased against `master`.

**Review by phatblat — APPROVED** [review]

**Comment by chris-araman** [comment]
Hrm... I'm not sure what the `build-test` check is waiting on here. It's already passed:
https://github.com/chris-araman/mas/actions/runs/696277174

**Comment by phatblat** [comment]
Oh, I think it's confused since you submitted this one from a fork. I'll override to get this big PR in before the other ones.

@chris-araman now that you're a contributor, you can create branches right on the `mas` repo which might work better.

**Pull request merged**